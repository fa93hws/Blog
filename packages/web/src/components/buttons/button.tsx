import * as React from 'react';
import * as classnames from 'classnames';
import { Link } from 'react-router-dom';
import * as Icons from 'components/icons/icons';
import { RipplableProps, withRipple, RippleColor } from './ripple/ripple';
import styles from './button.css';

type ButtonSize = 'medium' | 'large' | undefined;
type ButtonType = 'primary' | 'ghost';
type BaseButtonProps = {
  prefixIcon?: keyof typeof Icons;
  onClick?(e: React.MouseEvent<HTMLElement, MouseEvent>): void;
  onMouseDown?(e: React.MouseEvent<HTMLElement, MouseEvent>): void;
  circle?: boolean;
  children?: React.ReactNode;
  size?: ButtonSize;
  link?: string;
  type: ButtonType;
};

const sizeClassNameMap: Record<NonNullable<ButtonSize>, string> = {
  medium: styles.medium,
  large: styles.large,
};
const rippleColorMap: Record<ButtonType, RippleColor> = {
  primary: 'white',
  ghost: 'blue',
};

function getButtonClassName({
  size = 'medium',
  type,
  circle = false,
}: {
  size?: ButtonSize;
  type: ButtonType;
  circle?: boolean;
}) {
  return classnames(styles.baseButton, sizeClassNameMap[size], styles[type], {
    [styles.circle]: circle,
  });
}

function generateAttributes(props: BaseButtonProps & RipplableProps) {
  return {
    onClick: props.onClick,
    onMouseDown: props.onMouseDown,
    className: getButtonClassName({
      size: props.size,
      circle: props.circle,
      type: props.type,
    }),
  };
}
const ButtonChildren = React.memo((props: BaseButtonProps & RipplableProps) => {
  const PrefixIcon = props.prefixIcon == null ? null : Icons[props.prefixIcon];
  return (
    <>
      {PrefixIcon && <PrefixIcon size="small" />}
      {props.children && (
        <span
          className={classnames(styles.text, {
            [styles.withIcon]: PrefixIcon != null,
          })}
        >
          {props.children}
        </span>
      )}
      <div className={styles.rippleContainer}>
        <props.RippleSlot color={rippleColorMap[props.type]} />
      </div>
    </>
  );
});

const ButtonStateless = (props: BaseButtonProps & RipplableProps) => {
  const attributes = generateAttributes(props);
  return (
    <button {...attributes}>
      <ButtonChildren {...props} />
    </button>
  );
};
const LinkButtonStateless = (
  props: BaseButtonProps & RipplableProps & { link: string },
) => {
  const attributes = generateAttributes(props);
  if (props.link.startsWith('http')) {
    return (
      <a {...attributes} href={props.link}>
        <ButtonChildren {...props} />
      </a>
    );
  }
  return (
    <Link {...attributes} to={props.link}>
      <ButtonChildren {...props} />
    </Link>
  );
};

function createButton(type: ButtonType) {
  return React.memo((props: Omit<BaseButtonProps & RipplableProps, 'type'>) => {
    if (props.link == null) {
      return <ButtonStateless type={type} {...props} />;
    }
    return <LinkButtonStateless type={type} {...props} link={props.link} />;
  });
}

export const PrimaryButton = withRipple(createButton('primary'));
export const GhostButton = withRipple(createButton('ghost'));
