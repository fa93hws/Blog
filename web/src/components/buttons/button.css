@value gridBaseline from '~components/metrics.css';
@value colorBlue, colorBlueA250, colorBlueA400 from '~components/colors.css';
@value monoWhite from '~components/colors.css';
@value textSmall from '~components/typography/typography.css';

@value heightMedium: calc(gridBaseline * 4);
@value heightLarge: calc(gridBaseline * 5);

.baseButton {
  position: relative;
  display: inline-flex;
  align-items: center;
  justify-content: center;

  box-sizing: border-box;
  outline: 0;
  padding: 0 calc(gridBaseline * 2);
  border-radius: calc(gridBaseline * 0.5);
  cursor: pointer;

  transition-property: box-shadow, opacity;
  transition-duration: 500ms;
}

.rippleContainer {
  position: absolute;
  top: 0;
  right: 0;
  bottom: 0;
  left: 0;
  pointer-events: none;
  overflow: hidden;
  border-radius: calc(gridBaseline * 0.5);
}

/* colors */
.primary {
  background-color: colorBlue;
  border: 1px solid colorBlue;
  color: monoWhite;
  fill: monoWhite;
}

.primary:hover {
  opacity: 0.9;
}

.primary:active {
  opacity: 0.7;
}

.ghost {
  background-color: transparent;
  color: colorBlue;
  fill: colorBlue;
  border: none;
}

.ghost:hover {
  background-color: colorBlueA400;
}

.ghost:active {
  background-color: colorBlueA250;
}

/* shape */
.circle {
  border-radius: 50%;
}
.circle > .rippleContainer {
  border-radius: 50%;
}

/* size */
.text {
  composes: textSmall;
}

.text.withIcon {
  margin-left: gridBaseline;
}

.medium {
  height: heightMedium;
  min-width: calc(gridBaseline * 8);
}
.medium.circle {
  width: heightMedium;
  min-width: unset;
}

.large {
  height: heightLarge;
  min-width: calc(gridBaseline * 10);
}
.large.circle {
  width: heightLarge;
  min-width: unset;
}

/* box-shadow */
.primary {
  box-shadow: 0px 3px 1px -2px rgba(0, 0, 0, 0.2),
    0px 2px 2px 0px rgba(0, 0, 0, 0.14), 0px 1px 5px 0px rgba(0, 0, 0, 0.12);
}
.primary:hover {
  box-shadow: 0px 2px 4px -1px rgba(0, 0, 0, 0.2),
    0px 4px 5px 0px rgba(0, 0, 0, 0.14), 0px 1px 10px 0px rgba(0, 0, 0, 0.12);
}

.primary:active {
  box-shadow: 0px 5px 5px -3px rgba(0, 0, 0, 0.2),
    0px 8px 10px 1px rgba(0, 0, 0, 0.14), 0px 3px 14px 2px rgba(0, 0, 0, 0.12);
}
