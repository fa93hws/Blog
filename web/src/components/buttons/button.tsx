import * as React from 'react';
import * as classnames from 'classnames';
import { UnreachableException } from 'utils/exceptions/unreachable';
import * as Icons from 'components/icons/icons';
import { RipplableProps, withRipple, RippleColor } from './ripple/ripple';
import styles from './button.css';

type ButtonSize = 'medium' | 'large' | undefined;
type ButtonType = 'primary' | 'ghost';
type BaseButtonProps = {
  prefixIcon?: keyof typeof Icons;
  onClick?(e: React.MouseEvent<HTMLButtonElement, MouseEvent>): void;
  onMouseDown?(e: React.MouseEvent<HTMLButtonElement, MouseEvent>): void;
  circle?: boolean;
  children?: React.ReactNode;
  size?: ButtonSize;
  type: ButtonType;
};

function getSizeClassName(size: ButtonSize) {
  switch (size) {
    case undefined:
    case 'medium':
      return styles.medium;
    case 'large':
      return styles.large;
    default:
      throw new UnreachableException(size);
  }
}
const rippleColorMap: Record<ButtonType, RippleColor> = {
  primary: 'white',
  ghost: 'blue',
};

const ButtonStateless = (props: BaseButtonProps & RipplableProps) => {
  const PrefixIcon = props.prefixIcon == null ? null : Icons[props.prefixIcon];
  return (
    <button
      onClick={props.onClick}
      onMouseDown={props.onMouseDown}
      className={classnames(
        styles.baseButton,
        getSizeClassName(props.size),
        styles[props.type],
        {
          [styles.circle]: props.circle,
        },
      )}
    >
      {PrefixIcon && <PrefixIcon size="small" />}
      {props.children && (
        <span
          className={classnames(styles.text, {
            [styles.withIcon]: PrefixIcon != null,
          })}
        >
          {props.children}
        </span>
      )}
      <div className={styles.rippleContainer}>
        <props.RippleSlot color={rippleColorMap[props.type]} />
      </div>
    </button>
  );
};

function createButton(type: ButtonType) {
  return React.memo((props: Omit<BaseButtonProps & RipplableProps, 'type'>) => (
    <ButtonStateless type={type} {...props} />
  ));
}

export const PrimaryButton = withRipple(createButton('primary'));
export const GhostButton = withRipple(createButton('ghost'));
