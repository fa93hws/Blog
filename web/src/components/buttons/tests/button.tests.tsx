import * as React from 'react';
import { PrimaryButton, GhostButton } from '../button';

const ButtonMap = { PrimaryButton, GhostButton };
type TestCase = keyof typeof ButtonMap;
const testCases = Object.keys(ButtonMap) as TestCase[];

describe.each(testCases)('%s', (name: TestCase) => {
  const Button = ButtonMap[name];

  it('renders the default', () => {
    expect(<Button>Default</Button>).toMatchRenderedSnapshot();
  });

  it('renders the circle', () => {
    expect(<Button circle>Circle</Button>).toMatchRenderedSnapshot();
  });

  it('renders with icon', () => {
    expect(<Button prefixIcon="Github">Icon</Button>).toMatchRenderedSnapshot();
  });
});
