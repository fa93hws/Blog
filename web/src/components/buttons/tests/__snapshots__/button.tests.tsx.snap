// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`GhostButton renders the circle 1`] = `
<button
  class="baseButton medium ghost circle"
>
  <span
    class="text"
  >
    Circle
  </span>
  <div
    class="rippleContainer"
  />
</button>
`;

exports[`GhostButton renders the default 1`] = `
<button
  class="baseButton medium ghost"
>
  <span
    class="text"
  >
    Default
  </span>
  <div
    class="rippleContainer"
  />
</button>
`;

exports[`GhostButton renders with icon 1`] = `
<button
  class="baseButton medium ghost"
>
  <i
    aria-hidden="true"
    class="iconSpan"
  >
    <svg
      height="16"
      version="1.1"
      viewBox="0 0 16 16"
      width="16"
      xmlns="http://www.w3.org/2000/svg"
    >
      <g
        transform="scale(1)"
      >
        src/components/icons/assets/github.svg
      </g>
    </svg>
  </i>
  <span
    class="text withIcon"
  >
    Icon
  </span>
  <div
    class="rippleContainer"
  />
</button>
`;

exports[`PrimaryButton renders the circle 1`] = `
<button
  class="baseButton medium primary circle"
>
  <span
    class="text"
  >
    Circle
  </span>
  <div
    class="rippleContainer"
  />
</button>
`;

exports[`PrimaryButton renders the default 1`] = `
<button
  class="baseButton medium primary"
>
  <span
    class="text"
  >
    Default
  </span>
  <div
    class="rippleContainer"
  />
</button>
`;

exports[`PrimaryButton renders with icon 1`] = `
<button
  class="baseButton medium primary"
>
  <i
    aria-hidden="true"
    class="iconSpan"
  >
    <svg
      height="16"
      version="1.1"
      viewBox="0 0 16 16"
      width="16"
      xmlns="http://www.w3.org/2000/svg"
    >
      <g
        transform="scale(1)"
      >
        src/components/icons/assets/github.svg
      </g>
    </svg>
  </i>
  <span
    class="text withIcon"
  >
    Icon
  </span>
  <div
    class="rippleContainer"
  />
</button>
`;
